packages/components/src/theme/color-algorithms.ts

```diff
@@ -0,0 +1,126 @@
+/**
+ * External dependencies
+ */
+import { colord, extend } from 'colord';
+import a11yPlugin from 'colord/plugins/a11y';
+import namesPlugin from 'colord/plugins/names';
+
+/**
+ * WordPress dependencies
+ */
+import warning from '@wordpress/warning';
+
+/**
+ * Internal dependencies
+ */
+import type { ThemeInputValues, ThemeOutputValues } from './types';
+import { COLORS } from '../utils';
+
+extend( [ namesPlugin, a11yPlugin ] );
+
+export function generateThemeVariables(
+	inputs: ThemeInputValues
+): ThemeOutputValues {
+	validateInputs( inputs );
+
+	return {
+		colors: {
+			...generateAccentDependentColors( inputs.accent ),
+			...generateBackgroundDependentColors( inputs.background ),
+		},
+	};
+}
+
+export function validateInputs( inputs: ThemeInputValues ) {
+	for ( const [ key, value ] of Object.entries( inputs ) ) {
+		if ( typeof value !== 'undefined' && ! colord( value ).isValid() ) {
+			warning(
+				`wp.components.Theme: "${ value }" is not a valid color value for the '${ key }' prop.`
+			);
+		}
+	}
+
+	if ( inputs.background ) {
+		if (
+			! colord( inputs.background ).isReadable(
+				inputs.accent || COLORS.ui.theme
+			)
+		) {
+			warning(
+				`wp.components.Theme: The background color provided ("${ inputs.background }") does not have sufficient contrast against the accent color ("${ inputs.accent }").`
+			);
+		}
+
+		if (
+			! colord( inputs.background ).isReadable(
+				getForegroundForColor( inputs.background )
+			)
+		) {
+			warning(
+				`wp.components.Theme: The background color provided ("${ inputs.background }") does not have sufficient contrast against the standard foreground colors.`
+			);
+		}
+	}
+}
+
+function generateAccentDependentColors( accent?: string ) {
+	if ( ! accent ) return {};
+
+	return {
+		accent,
+		accentDarker10: colord( accent ).darken( 0.1 ).toHex(),
+		accentDarker20: colord( accent ).darken( 0.2 ).toHex(),
+		accentInverted: getForegroundForColor( accent ),
+	};
+}
+
+function generateBackgroundDependentColors( background?: string ) {
+	if ( ! background ) return {};
+
+	const foreground = getForegroundForColor( background );
+
+	return {
+		background,
+		foreground,
+		foregroundInverted: getForegroundForColor( foreground ),
+		gray: generateShades( background, foreground ),
+	};
+}
+
+function getForegroundForColor( color: string ) {
+	return colord( color ).isDark() ? COLORS.white : COLORS.gray[ 900 ];
+}
+
+export function generateShades( background: string, foreground: string ) {
```

Comment:
That could be a good idea, although it's hard to say _how much_ until we convert a few more components and see how it feels. Something with borders and help text or something.

In a sense, the consumer already has control over this because the grayscale colors will simply maintain the saturation of the provided background color... which I _think_ should be the most harmonious? I'm not sure yet.

I added this to the tracking issue so we can revisit once we have a few more components converted.